File: App/SharedProfileSyncCoordinator.swift
import Domain
import Foundation
import Store
import SwiftData
import Sync

/// Doc 14 « Profils partagés », phase 2 — pousse le résumé d'une partie tout juste conclue vers
/// chaque participant lié à l'installation d'un ami, et récupère les résumés que d'autres ont
/// poussés vers l'une de mes propres fiches liées. Déclenché par `CaCompteApp` (lancement et
/// retour au premier plan), pas par un minuteur propre — même discipline que
/// `MatchConnectionCoordinator`, plutôt qu'un nouveau système de synchronisation.
@MainActor
final class SharedProfileSyncCoordinator {
    static let shared = SharedProfileSyncCoordinator()

    private let transport = SharedProfileTransport()
    private var isSyncing = false

    private init() {}

    func sync(context: ModelContext) async {
        guard !isSyncing else { return }
        isSyncing = true
        defer { isSyncing = false }
        await push(context: context)
        await pull(context: context)
    }

    /// Un résumé par partie conclue avec au moins un participant lié — jamais retenté
    /// indéfiniment pour une partie qui n'en a plus (fiche déliée entre-temps, par exemple).
    private func push(context: ModelContext) async {
        let repository = MatchRepository(context: context)
        guard let pendingMatches = try? repository.matchesPendingSharedProfileSync(), !pendingMatches.isEmpty else { return }

        for match in pendingMatches {
            let linkedIDs = Array(Set(match.participants.compactMap { $0.player?.sharedProfileID }))
            let standings = match.participants.compactMap { participant -> SharedMatchSummaryPayload.Entry? in
                guard let rank = participant.finalRank, let score = participant.finalScore else { return nil }
                return SharedMatchSummaryPayload.Entry(
                    sharedProfileID: participant.player?.sharedProfileID,
                    nickname: participant.nicknameSnapshot,
                    avatarKind: participant.avatarKindSnapshot,
                    avatarValue: participant.avatarValueSnapshot,
                    paletteID: participant.paletteIDSnapshot,
                    rank: rank,
                    score: score
                )
            }

            guard !linkedIDs.isEmpty, !standings.isEmpty else {
                try? repository.markSharedProfileSyncComplete(match)
                continue
            }

            let payload = SharedMatchSummaryPayload(
                gameID: match.gameID,
                rulesVersion: match.rulesVersion,
                playedAt: match.endedAt ?? match.startedAt,
                standings: standings
            )
            let rows = linkedIDs.map { SharedMatchSummaryRow(matchID: match.id, sharedProfileID: $0, payload: payload) }

            do {
                try await transport.push(rows)
                try? repository.markSharedProfileSyncComplete(match)
            } catch {
                // Échec silencieux : `pendingSharedProfileSync` reste `true`, nouvelle tentative
                // au prochain retour au premier plan (même patron que `MatchConnectionCoordinator`).
            }
        }
    }

    /// `materializeSharedSummary` est déjà un no-op si cette partie est connue localement (c'est
    /// cet appareil qui l'a jouée et poussée) — reste alors juste à nettoyer la boîte aux lettres.
    private func pull(context: ModelContext) async {
        let playerRepository = PlayerRepository(context: context)
        let matchRepository = MatchRepository(context: context)
        guard let linkedIDs = try? playerRepository.allSharedProfileIDs(), !linkedIDs.isEmpty else { return }
        guard let rows = try? await transport.fetchPending(for: linkedIDs) else { return }

        for row in rows {
            try? matchRepository.materializeSharedSummary(row.payload, matchID: row.matchID)
            try? await transport.delete(matchID: row.matchID, sharedProfileID: row.sharedProfileID)
        }
    }
}

File: android/app/src/main/kotlin/com/cacompte/app/profilesharing/SharedProfileSyncCoordinator.kt
package com.cacompte.app.profilesharing

import com.cacompte.domain.model.SharedMatchSummaryPayload
import com.cacompte.store.MatchRepository
import com.cacompte.store.PlayerRepository
import com.cacompte.sync.SharedMatchSummaryRow
import com.cacompte.sync.SharedProfileTransport
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.first

/**
 * Miroir de `SharedProfileSyncCoordinator.swift` (doc 14, phase 2) — pousse le résumé d'une
 * partie tout juste conclue vers chaque participant lié à l'installation d'un ami, et récupère
 * les résumés que d'autres ont poussés vers l'une de mes propres fiches liées. Déclenché au
 * lancement et au retour au premier plan (voir `com.cacompte.app.CaCompteApplication`), pas par
 * un minuteur propre — même discipline que `com.cacompte.app.livesync.MatchConnectionCoordinator`.
 */
class SharedProfileSyncCoordinator(
    private val matchRepository: MatchRepository,
    private val playerRepository: PlayerRepository,
) {
    private val transport = SharedProfileTransport()
    private var isSyncing = false

    suspend fun sync() {
        if (isSyncing) return
        isSyncing = true
        try {
            push()
            pull()
        } finally {
            isSyncing = false
        }
    }

    /** Un résumé par partie conclue avec au moins un participant lié — jamais retenté
     * indéfiniment pour une partie qui n'en a plus (fiche déliée entre-temps, par exemple). */
    private suspend fun push() {
        val pending = matchRepository.matchesPendingSharedProfileSync()
        if (pending.isEmpty()) return
        val playersByID = playerRepository.observeAll().first().associateBy { it.id }

        for (match in pending) {
            val participants = matchRepository.participants(match.id)
            val linkedIDs =
                participants
                    .mapNotNull { participant -> participant.playerId?.let { playersByID[it]?.sharedProfileID } }
                    .toSet()
            val standings =
                participants.mapNotNull { participant ->
                    val rank = participant.finalRank ?: return@mapNotNull null
                    val score = participant.finalScore ?: return@mapNotNull null
                    SharedMatchSummaryPayload.Entry(
                        sharedProfileID = participant.playerId?.let { playersByID[it]?.sharedProfileID },
                        nickname = participant.nicknameSnapshot,
                        avatarKind = participant.avatarKindSnapshot,
                        avatarValue = participant.avatarValueSnapshot,
                        paletteID = participant.paletteIDSnapshot,
                        rank = rank,
                        score = score,
                    )
                }

            if (linkedIDs.isEmpty() || standings.isEmpty()) {
                matchRepository.markSharedProfileSyncComplete(match)
                continue
            }

            val payload =
                SharedMatchSummaryPayload(
                    gameID = match.gameID,
                    rulesVersion = match.rulesVersion,
                    playedAt = match.endedAt ?: match.startedAt,
                    standings = standings,
                )
            val rows =
                linkedIDs.map {
                    SharedMatchSummaryRow(
                        matchID = match.id,
                        sharedProfileID = it,
                        payload = payload,
                    )
                }

            try {
                transport.push(rows)
                matchRepository.markSharedProfileSyncComplete(match)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Exception) {
                // Échec silencieux : pendingSharedProfileSync reste true, nouvelle tentative au
                // prochain déclenchement (même patron que MatchConnectionCoordinator).
            }
        }
    }

    /** `materializeSharedSummary` est déjà un no-op si cette partie est connue localement (c'est
     * cet appareil qui l'a jouée et poussée). */
    private suspend fun pull() {
        val linkedIDs = playerRepository.allSharedProfileIDs()
        if (linkedIDs.isEmpty()) return
        val rows =
            try {
                transport.fetchPending(linkedIDs)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Exception) {
                return
            }

        val myOwnID = playerRepository.myOwnSharedPlayer()?.sharedProfileID

        for (row in rows) {
            val isMyOwnIdentity = row.sharedProfileID == myOwnID
            // Doc 14, phase 4 — suivre un ami donne accès à *toutes* ses parties seulement si
            // j'y étais moi-même (mon propre identifiant apparaît alors parmi les *autres*
            // participants du résumé) ; ma propre fiche partagée, elle, reçoit tout sans filtre.
            val isRelevant = isMyOwnIdentity || row.payload.standings.any { it.sharedProfileID == myOwnID }
            if (isRelevant) {
                matchRepository.materializeSharedSummary(row.payload, row.matchID)
            }

            // Doc 14, phase 4 — seul l'appareil qui fait autorité sur cet identifiant (le sien)
            // nettoie la boîte aux lettres ; plusieurs amis peuvent suivre la même personne, et
            // supprimer après chaque lecture ferait perdre la partie aux autres.
            if (isMyOwnIdentity) {
                try {
                    transport.delete(row.matchID, row.sharedProfileID)
                } catch (cancellation: CancellationException) {
                    throw cancellation
                } catch (error: Exception) {
                    // Best-effort — une ligne non nettoyée sera relue (idempotent) puis purgée côté serveur.
                }
            }
        }
    }
}
